sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/sycl_ext_oneapi_bindless_images.asciidoc

```diff
@@ -0,0 +1,541 @@
+top of it.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification. An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_BINDLESS_IMAGES` to one of the values defined in the
+table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[frame="none",options="header"]
+|======================
+|Value |Description
+|202302 |Initial version of this extension
+|======================
+
+=== Image descriptor
+
+```cpp
+namespace sycl::ext::oneapi {
+
+struct image_descriptor {
+  unsigned int width;
+  unsigned int height;
+  unsigned int depth;
+  image_channel_type channel_type;
+  image_channel_order channel_order;
+  unsigned int row_pitch;
+};
+
+}
+```
+
+The image descriptor represents the image dimensions, channel type, channel 
+order, and row pitch.
+
+Note that `image_channel_type` and `image_channel_order` existed in SYCL 1.2.1,
+but were removed in SYCL 2020 in favor of a single, unified enum class.
+We propose separating them again to enable better flexibility
+and to avoid combinatorial complexity.
+
+=== Allocating image memory
+
+```cpp
+namespace sycl::ext::oneapi {
+
+image_mem_handle allocate_image(const context& syclContext, 
+                                image_descriptor desc);
+
+void free_image(const context& syclContext, void *memory_handle);
+
+}
+```
+
+The process of creating an image is two-fold:
+allocate an image, then create an image handle from the allocation.
+
+The first step is to allocate memory for an image.
+`allocate_image` can be used for this,
+as it allocates memory based on the requirements outlined by the image 
+descriptor. The function returns a handle to the newly allocated memory.
+Note that the image is an opaque type, so the handle cannot be dereferenced on 
+the host. The layout of the allocated memory is backend-specific, and may be an 
+optimized layout, e.g. tile swizzle patterns.
+
+After we're done with the image, we need to cleanup the memory using 
+`free_image`.
+
+_In the DPC++ CUDA backend, these will allocate/deallocate a `CUarray` type._
+
+An alternative is to allocate memory using device USM allocations.
+In addition to the existing USM allocation functions, a new USM pitched 
+allocation function is proposed.
+
+```cpp
+namespace sycl::ext::oneapi {
+
+void *pitched_alloc_device(size_t* ret_pitch, size_t width_in_bytes, 
+                           size_t height unsigned int element_size_bytes, 
+                           const queue &queue);
+
+}
+```
+
+This function will allocate a memory region aimed to be used for 
+two-dimensional images.
+It places appropriate padding at the end of image rows and returns the pitch 
+value used.
```

Comment:
> The function `allocate_image_mem` above uses `image_channel_type` (I think) to define the element size, while this API takes `element_size_bytes`. Why the difference?

`pitched_alloc_device` takes `element_size_bytes` instead of the image descriptor as the pitched allocation might be used outside of its application for images. The API was designed to be similar to how CUDA goes about allocating pitched memory.

The pitched allocation returns a pitch value chosen by the backend driver, we can have `pitched_alloc_device` take a non-const `image_descriptor` reference, and the pitch field of the descriptor could be populated. This way we would couple any pitched allocation with our images. 

> Is there no need for `image_channel_order` in this API?

This API was made for more general memory allocation, and not necessarily just image memory allocation. If we switch to passing the `image_descriptor` instead, we would use `image_channel_order` and `image_channel_type` along with the `width` and `height` to determine how to allocate memory.

> With `allocate_image_mem`, the row pitch is specified by the caller in `image_descriptor`, whereas in this API the row pitch is returned. Why the difference

`allocate_image_mem` doesn't allocate pitched memory, and the pitch value in the descriptor is not necessary for that function to work (it's ignored). I agree that this is a bit unclear and clunky. We should specify that `pitched_alloc_device` will allocate pitched memory and populate the passed `image_descriptor` with the appropriate pitch value, whereas with `allocate_image_mem` the pitch is not relevant and should be zero.

The reason that pitch is not always relevant is due to the backend type used to hold the image memory. In the CUDA backend, for non-USM images, we use a `CUarray`. When creating an image from `CUarray` we don't need to provide a pitch value. For USM images, however, we need to use a different resource type (`CU_RESOURCE_TYPE_PITCH2D` for 2D images), and pass memory backed by a `CUdeviceptr` along with a pitch value.

We might want to keep the current `pitched_alloc_device` API in case a user does not have the `image_descriptor` at hand, or wants to allocate pitched memory for uses outside of images. But having the option to pass the `image_descriptor` instead would be helpful.

I agree with the other points on the API and memory access restrictions, we will update the proposal appropriately.